glean-core/src/metrics/object.rs

```diff
@@ -0,0 +1,155 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::sync::Arc;
+
+use crate::common_metric_data::CommonMetricDataInternal;
+use crate::error_recording::{record_error, test_get_num_recorded_errors, ErrorType};
+use crate::metrics::Metric;
+use crate::metrics::MetricType;
+use crate::storage::StorageManager;
+use crate::CommonMetricData;
+use crate::Glean;
+
+use serde_json::Value as JsonValue;
+
+/// A counter metric.
+///
+/// Used to count things.
+/// The value can only be incremented, not decremented.
+#[derive(Clone, Debug)]
+pub struct ObjectMetric {
+    meta: Arc<CommonMetricDataInternal>,
+}
+
+impl MetricType for ObjectMetric {
+    fn meta(&self) -> &CommonMetricDataInternal {
+        &self.meta
+    }
+}
+
+// IMPORTANT:
+//
+// When changing this implementation, make sure all the operations are
+// also declared in the related trait in `../traits/`.
+impl ObjectMetric {
+    /// Creates a new counter metric.
+    pub fn new(meta: CommonMetricData) -> Self {
+        Self {
+            meta: Arc::new(meta.into()),
+        }
+    }
+
+    /// Sets to the specified structure.
+    ///
+    /// # Arguments
+    ///
+    /// * `glean` - the Glean instance this metric belongs to.
+    /// * `value` - the value to set.
+    #[doc(hidden)]
+    pub fn set_sync(&self, glean: &Glean, value: JsonValue) {
+        let value = Metric::Object(serde_json::to_string(&value).unwrap());
+        dbg!(&value);
+        glean.storage().record(glean, &self.meta, &value)
+    }
+
+    /// Sets to the specified structure.
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - the value to set.
+    ///
+    /// # Returns
+    ///
+    /// The stored value or `None` if nothing stored.
+    pub fn set(&self, value: JsonValue) {
+        let metric = self.clone();
+        crate::launch_with_glean(move |glean| metric.set_sync(glean, value))
+    }
+
+    /// Record an `InvalidValue` error for this metric.
+    ///
+    /// Only to be used by the RLB.
+    pub fn record_invalid_value(&self) {
```

Comment:
Since the error message is never not about the schema, the name of this fn should probably be `record_schema_error` or something